sympy/geometry/ellipse.py

```diff
@@ -695,25 +693,50 @@ def is_tangent(self, o):
         True
 
         """
-        inter = None
-        if isinstance(o, Ellipse):
-            inter = self.intersection(o)
-            if isinstance(inter, Ellipse):
+        if isinstance(o, Point2D):
+            return False
+        elif isinstance(o, Ellipse):
+            intersect = self.intersection(o)
+            if isinstance(intersect, Ellipse):
+                return True
+            elif intersect:
+                if all([(self.tangent_lines(i)[0]).equals((o.tangent_lines(i)[0])) for i in intersect]):
+                    return True
+                else:
+                    return False
+            else:
+                return False
+        elif isinstance(o, Line2D):
+            if len(self.intersection(o)) == 1:
+                return True
+            else:
                 return False
-            return (inter is not None and len(inter) == 1
-                    and isinstance(inter[0], Point))
-        elif isinstance(o, LinearEntity):
-            inter = self.intersection(o)
-            if inter is not None and len(inter) == 1:
-                return inter[0] in o
+        elif isinstance(o, Ray2D):
+            intersect = self.intersection(o)
+            if len(intersect) == 1:
+                if intersect[0] != o.source and not self.encloses_point(o.source):
+                    return True
+                else:
+                    return False
             else:
                 return False
-        elif isinstance(o, Polygon):
-            c = 0
-            for seg in o.sides:
-                inter = self.intersection(seg)
-                c += len([True for point in inter if point in seg])
-            return c == 1
+        elif isinstance(o, (Segment2D, Polygon)):
+            any_intersect = False
+            segments = o.sides if isinstance(o, Polygon) else [o]
+            for segment in segments:
```

Comment:
Ah, I see my idea won't work, but I was trying to suggest something more recursive.